tinygraph/tinygraph.c

```diff
@@ -506,3 +507,334 @@ bool tinygraph_reorder(
   // allocation. For now this is always returning true.
   return true;
 }
+
+
+// The purpose of the dijkstra context is to cache state
+// like the distance and parents array, so that we don't
+// have to allocate memory for every new s-t search.
+typedef struct tinygraph_dijkstra {
+  uint32_t s;
+  uint32_t t;
+
+  uint32_t* dist;
+  uint32_t* parent;
+  tinygraph_array_s path;
+
+  const uint16_t* weight;
+
+  tinygraph_const_s graph;
+  tinygraph_bitset_s seen;
+  tinygraph_heap_s heap;
+} tinygraph_dijkstra;
+
+
+static inline void tinygraph_dijkstra_clear(tinygraph_dijkstra_s ctx) {
+  ctx->s = UINT32_MAX;
+  ctx->t = UINT32_MAX;
+
+  const uint32_t n = tinygraph_get_num_nodes(ctx->graph);
+
+  for (uint32_t i = 0; i < n; ++i) {
+    ctx->dist[i] = UINT32_MAX;
+  }
+
+  for (uint32_t i = 0; i < n; ++i) {
+    ctx->parent[i] = i;
+  }
+
+  tinygraph_bitset_clear(ctx->seen);
+  tinygraph_heap_clear(ctx->heap);
+}
+
+
+tinygraph_dijkstra_s tinygraph_dijkstra_construct(tinygraph_const_s graph, const uint16_t* weights) {
+  TINYGRAPH_ASSERT(graph);
+  TINYGRAPH_ASSERT(!tinygraph_is_empty(graph));
+
+  tinygraph_dijkstra *out = malloc(sizeof(tinygraph_dijkstra));
+
+  if (!out) {
+    return NULL;
+  }
+
+  const uint32_t n = tinygraph_get_num_nodes(graph);
+
+  tinygraph_bitset_s seen = tinygraph_bitset_construct(n);
+
+  if (!seen) {
+    free(out);
+
+    return NULL;
+  }
+
+  tinygraph_heap_s heap = tinygraph_heap_construct();
+
+  if (!heap) {
+    tinygraph_bitset_destruct(seen);
+    free(out);
+
+    return NULL;
+  }
+
+  uint32_t* dist = malloc(n * sizeof(uint32_t));
+
+  if (!dist) {
+    tinygraph_heap_destruct(heap);
+    tinygraph_bitset_destruct(seen);
+    free(out);
+
+    return NULL;
+  }
+
+  uint32_t* parent = malloc(n * sizeof(uint32_t));
+
+  if (!parent) {
+    free(dist);
+    tinygraph_heap_destruct(heap);
+    tinygraph_bitset_destruct(seen);
+    free(out);
+
+    return NULL;
+  }
+
+  tinygraph_array_s path = tinygraph_array_construct(0);
+
+  if (!path) {
+    free(parent);
+    free(dist);
+    tinygraph_heap_destruct(heap);
+    tinygraph_bitset_destruct(seen);
+    free(out);
+
+    return NULL;
+  }
+
+  *out = (tinygraph_dijkstra){
+    .s = UINT32_MAX,
+    .t = UINT32_MAX,
+    .dist = dist,
+    .parent = parent,
+    .path = path,
+    .weight = weights,
+    .graph = graph,
+    .seen = seen,
+    .heap = heap,
+  };
+
+  // Resets the internal state e.g. sets
+  // self-loops for the parents array
+  tinygraph_dijkstra_clear(out);
+
+  return out;
+}
+
+
+void tinygraph_dijkstra_destruct(tinygraph_dijkstra * const ctx) {
+  if (!ctx) {
+    return;
+  }
+
+  ctx->s = UINT32_MAX;
+  ctx->t = UINT32_MAX;
+
+  free(ctx->dist);
+  free(ctx->parent);
+
+  tinygraph_array_destruct(ctx->path);
+  tinygraph_heap_destruct(ctx->heap);
+  tinygraph_bitset_destruct(ctx->seen);
+
+  free(ctx);
+}
+
+
+static inline uint32_t tinygraph_saturated_add_u32(uint32_t a, uint32_t b) {
```

Comment:
I'm curious if this will be optimized to `__builtin_uadd_overflow`. :thinking: Did you check that by any chance?

Reply:
It compiles to an addition and a conditional move based on the carry flag from the addition

https://godbolt.org/z/3crhPsxo8

Did a quick perf run (check the makefile) and most time spent is in the binary heap, all the computation here is pretty much for free while we're chasing through memory.